Discern.lic

Rhayle opened the pull request:
A script to be added to autorun, if desired, capable of running discern and populating the discern array before executing other scripts.

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+custom_require.call(%w[common common-travel events spellmonitor drinfomon])
```
Please add a documentation header so we can put additional information on elanthipedia

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+        UserVars.discerns = {}
+    end
+    echo 'Reading spell list'
+    buffspelllist = settings.buff_spells
```
What do you think about `buff_spell_list`? (and similar for the next two vars).

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+    buffspelllist = settings.buff_spells
+    offspelllist =  settings.offensive_spells
+    ctspelllist = settings.training_spells
+     buffspelllist.each {|x| if (x[1]['use_auto_mana'])
```
What do you think about extracting the `if` into a chained method like `buff_spell_list.select{|x| (x[1]['use_auto_mana'])}.each {|x| ...}`?

rpherbig replied:
This also applies to the other blocks

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+    offspelllist =  settings.offensive_spells
+    ctspelllist = settings.training_spells
+     buffspelllist.each {|x| if (x[1]['use_auto_mana'])
+        @abbrev = x[1]['abbrev'] ? x[1]['abbrev'] : x[0]
```
Why is this an instance var (`@`)? Could it be local instead (`abbrev`)?

rpherbig replied:
This also applies to the other blocks

rpherbig commented:
@jandersson do you have a few minutes to review this?

rpherbig commented on astrology.lic:
```diff
@@ -46,14 +46,21 @@ class Astrology
     buffs.reject! { |spell| spell.eql?('Read the Ripples') }
+    
+    buffs.keys.each{ |key| 
+        if(buffs[key]['use_auto_mana'])
```
Same comment here about method chaining, something like `buffs.keys.select{|key| buffs[key]['use_auto_mana']}.each{...}`

rpherbig replied:
Or maybe `buffs.select{|key, _v| buffs[key]['use_auto_mana']}.each_value{|value| check_discern(value,settings)}`

Since this is an array, you don't really need to `.keys` and then later `buffs[key]`.

rpherbig commented on buff.lic:
```diff
@@ -30,6 +30,12 @@ class Waggle
     elsif DRStats.thief?
       start_khris(spells, settings)
     else
+      spells.keys.each{ |key| 
```
Same comment here about method chaining

jandersson commented on astrology.lic:
```diff
@@ -46,14 +46,21 @@ class Astrology
                      .first
     buffs.reject! { |spell| spell.eql?('Read the Ripples') }
+    
+    buffs.keys.each{ |key| 
```
I think rubocop will change this to `each_key`. Would you mind running rubocop on this? if you do `bundle install` while in the repository root directory it will install rubocop. Then you can do `mv astrology.lic astrology.rb` run rubocop in autocorrect mode, `rubocop -a astrology.rb` then `mv astrology.rb astrology.lic`. Alternatively you can do `rubocop -a` which will autocorrect all scripts.

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+
+    def initialize()
+    settings = get_settings
+    if(settings.force_discerns)
```
Is this a new setting? It feels more like it should be an argument to the script...

jandersson replied:
If its a new setting, it needs to be added to base.yaml

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,46 @@
+    echo 'Reading spell list'
+    buffspelllist = settings.buff_spells
+    offspelllist =  settings.offensive_spells
+    ctspelllist = settings.training_spells
```
Do you need to store each of these spell lists into a local var? What do you think about something like `settings.training_spells.each {|x| ...}` instead?

Rhayle commented:
I cannot get it to run rubocop.

jandersson commented on discern.lic:
```diff
@@ -0,0 +1,40 @@
+      echo " Current Discern data valid until '#{UserVars.discerns[@abbrev]['time_stamp'] + 24 * 60 * 60}'"
+    end
+end
+end
```
This might throw an error, I can find a matching statement that begins this

jandersson commented:
Youll want to expand all the previous comments by clicking 'show outdated' as rubocop seems to have beautified your code quite a bit :)

Rhayle commented:
I finally got rubocop to run.

rpherbig commented on astrology.lic:
```diff
@@ -54,6 +54,11 @@ class Astrology
                      .first
     buffs.reject! { |spell| spell.eql?('Read the Ripples') }
+
+    buffs.keys.each do |key|
```
What do you think about using the hash directly instead of `.keys`? Something like this:
```ruby
    buffs.each do |key, value|
      check_discern(value, settings) if value['use_auto_mana']
    end
```

rpherbig replied:
What do you think about using method chaining for the `if`? Something like this:
```ruby
    buffs.select { |_key, value| value['use_auto_mana'] }
         .each { |_key, value| check_discern(value, settings) }
```

Rhayle replied:
can you explain this syntax with the if statement. It looks to me like that would be applied to everything following the statement than what is before the statement.

rpherbig replied:
The original code iterates over every item, but only runs `check_discern` if a condition is met.

The `.select` syntax only keeps items for which the condition is met, then runs `check_discern` over everything that was kept.

See https://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-select for the technical details.

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,42 @@
+
+  def initialize
+    settings = get_settings
+    UserVars.discerns = {} if settings.force_discerns
```
I'm not sure what I think of `force_discerns` being a setting. That implies that you want to re-discern every time this script is run. Is that a common use case?

This feels more like a script argument to me, so you can specify it (or not) differently each time the script is run. What do you think?

Rhayle replied:
grumble no I thought I removed the force setting it was only for me to test with.  i figure you can run ;vars delete discerns instead of this setting

rpherbig commented on discern.lic:
```diff
@@ -0,0 +1,42 @@
+    UserVars.discerns = {} if settings.force_discerns
+    echo 'Reading spell list'
+
+    settings.buff_spells.each do |x|
```
I have a similar thought here about method chaining, something like:
```ruby
settings.buff_spells
    .select { |x| x[1]['use_auto_mana'] }
    .each do |x| ...
```

What do you think?

Rhayle commented:
ok. See what you think of this. I've made requested changes and fixed errors. The .select .each did not work in part of the discern scripts so I corrected to use key value

rpherbig commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.0.0-p647
+2.4
```
Was this supposed to be checked in?

Rhayle replied:
no. sorry. I thought I had it in the ignores.
How do I remove it?

rpherbig commented:
Sorry for the confusion there. Thank you so much for your contribution @Rhayle !
